Add support for OpenVox beaker jobs

bastelfreak opened the pull request.

bastelfreak commented:
I know that this isn't ready yet. The focus is on creating beaker jobs for puppet and openvox packages. unit tests will continue with the `puppet` gem.

bastelfreak commented:
>  want to do similar things with unit tests to test with agent gems

Yes. But since we don't have the gem published at the moment, it didn't have a high priority for me.

> Compatible solution is to just expand that result, even if the name includes puppet

That's my idea. I want to expand it, if `openvox` is listed in the requirements. In that way it isn't a breaking change. And if individual modules don't want to/cannot support Perforce packages anymore, we drop the `puppet` key from the requirements. And if there's a situation where we don't want to use the Perforce packages in any module, we could just reduce the result again.

ekohl commented:
Right, so let's start 2 variable that are generic: `puppet_metadata_unix_matrix` and `puppet_metadata_acceptance_matrix`. Those generate a GHA compatible matrix based on metadata. How we exactly define is is an implementation detail. Then in `gha-puppet` we start using the new matrices.

On the implementation details, for acceptance I'd say it os*requirement*versions. So centos 9 * {puppet,openvox} * 8 for example. It may be more complex, but that's up to this gem to decide. Similar for unit tests

ekohl commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -17,8 +17,9 @@ def initialize(metadata, options)
     def outputs(at = nil)
       {
         puppet_major_versions: puppet_major_versions,
```
This is a leftover from old matrix testing we no longer use. I'd lean to deprecating this instead of also adding `openvox_major_versions`.

bastelfreak replied:
Should we already remove it or only deprecate it? We already merged some breaking changes, so the next release is a major one anyway: https://github.com/voxpupuli/puppet_metadata/pull/166

ekohl commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -49,15 +62,19 @@ def puppet_unit_test_matrix
 
-    def beaker_os_releases(at = nil)
-      majors = puppet_major_versions
+    def beaker_os_releases(at = nil, collection)
```
The name `collection` isn't  something I entirely agree with. Previously we have Puppet PC1 as a collection, later `puppet5`, `puppet6`, etc. So IMHO `collection` isn't quite accurate.

Question is, what is accurate. In `metadata.json` it's listed under `requirements` so I'd play with something like `for_requirement` because I don't know a good word for the thing that satisfies the requirement.

ekohl replied:
Thinking more about it now: `requirement_name` could work as well

bastelfreak replied:
`requirement_name` sounds good to me

ekohl commented on lib/puppet_metadata/metadata.rb:
```diff
@@ -137,20 +137,28 @@ def satisfies_requirement?(name, version)
+      # raise Exception, "No #{collection} requirement found" unless requirement
+      return [] unless requirement
 
       # Current latest major is 7. It is highly recommended that modules
       # actually specify exact bounds, but this prevents an infinite loop.
       end_major = (requirement.end == SemanticPuppet::Version::MAX) ? 7 : requirement.end.major
```
This looks out of date, which is a problem with hardcoding magic numbers ...

bastelfreak replied:
probably not worth the dependency, but we could depend on rspec-puppet-facts, which ships https://github.com/voxpupuli/rspec-puppet-facts/blob/master/ext/puppet_agent_components.json

smortex replied:
I would really not mind to see hard failures when an upper bound is not set.

bastelfreak replied:
as a quick fix, I raised https://github.com/voxpupuli/puppet_metadata/pull/167

ekohl commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -89,17 +105,25 @@ def beaker_os_releases(at = nil)
     end
 
     def puppet_beaker_test_matrix(at)
+      beaker_test_matrix(at, 'puppet')
+    end
+
+    def openvox_beaker_test_matrix(at)
+      beaker_test_matrix(at, 'openvox')
```
And following up on the above discussion:
```suggestion
      beaker_test_matrix(at, 'puppet') + beaker_test_matrix(at, 'openvox')
```

You can even keep them alphabetically:
```suggestion
      beaker_test_matrix(at, 'openvox') + beaker_test_matrix(at, 'puppet')
```

ekohl commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -89,15 +105,21 @@ def beaker_os_releases(at = nil)
 
-      beaker_os_releases(at) do |os, release, puppet_version|
-        next if puppet_version_below_minimum?(puppet_version[:value])
+      beaker_os_releases(at, requirement_name) do |os, release, puppet_version|
```
This uses `puppet` variable naming quite often. What is a good way to describe either Puppet or OpenVox? `implementation` feels long and generic, but it's the best I can come up with.

bastelfreak replied:
switched to `implementation`

ekohl commented on lib/puppet_metadata/metadata.rb:
```diff
@@ -137,20 +137,28 @@ def satisfies_requirement?(name, version)
-      raise Exception, 'No Puppet requirement found' unless requirement
+    def major_versions(requirement_name)
+      requirement = requirements[requirement_name]
+      # don't raise an exception
+      # raise Exception, "No #{requirement_name} requirement found" unless requirement
```
I'm not a fan of commented code. I wonder if we should separate out the concerns so you can do the exception handling.

Thinking out loud about APIs:
```ruby
def puppet_major_versions
  requirement = requirements['puppet']
  return [] unless requirement

  major_version_for_requirement(requirement)
end

def puppet_major_versions!
  requirement = requirements['puppet']
  raise Exception, 'No Puppet requirement found' unless requirement

  major_version_for_requirement(requirement)
end

private

def major_version_for_requirement(requirement)
end
```

Not that you need `puppet_major_versions!` in the current iteration, but it's for an example.

Then taking that a step further, you can generalize it:
```ruby
def major_versions(requirement_name)
  requirement = requirements[requirement_name]
  return [] unless requirement

  major_version_for_requirement(requirement)
end

def major_versions!(requirement_name)
  requirement = requirements[requirement_name]
  raise Exception, "No '#{requirement_name}' requirement found' unless requirement

  major_version_for_requirement(requirement)
end

private

def major_version_for_requirement(requirement)
end
```

I like that a lot more because it keeps the `PuppetMetadata::Metadata` class very generic without naming individual requirements. We can accept that `PuppetMetadata::GithubActions` is a lot more specific while keeping this metadata abstraction as pure as possible.

The one thing that isn't pure is the missing upper bound handling. Technically that's already sort of broken, but in theory it could use some constant that's like this:
```ruby
REQUIREMENT_UPPER_BOUND_DEFAULTS = {
  'openvox' => 8,
  'puppet' => 8,
}
```

But I don't think `requirement` has a name so I'd postpone looking at it and keep the current implementation.
```

bastelfreak replied:
The chances are so so high that we need to drop puppet anyways in the near future, because of the unacceptable EULA, so let's postpone `REQUIREMENT_UPPER_BOUND_DEFAULTS` and see what the future brings. I implemented the rest.

ekohl commented on lib/puppet_metadata/aio.rb:
In this file `BUILDS` refers to what Puppet has AIO builds for. Should that have an extra dimension of `requirement_name`?

bastelfreak replied:
we probably need that in the future. At the moment it's not required I think, so I would like to make that an issue for future-bastelfreak.

ekohl reviewed (commented):
A lot of these things are me thinking out loud. I'd welcome counter arguments.

bastelfreak commented on lib/puppet_metadata/aio.rb:
```diff
@@ -42,6 +42,11 @@ class AIO
+    OPENVOX_RUBY_VERSIONS = {
+      7 => '2.7',
+      8 => '3.2',
+    }.freeze
```
added it

bastelfreak commented on lib/puppet_metadata/metadata.rb:
```diff
@@ -137,20 +137,28 @@ def satisfies_requirement?(name, version)
+      major_versions('openvox')
+    end
+
     # @return [Array[Integer]] Supported major Puppet versions
     # @see #requirements
```
I wanted to add them when we actually know which methods we will have.

bastelfreak commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -48,15 +60,19 @@ def puppet_unit_test_matrix
+                 puppet_major_versions
+               when 'openvox'
+                 openvox_major_versions
+               end
```
rubocop preferred `raise StandardError`, but I think that's still okay.

bastelfreak commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -89,15 +105,21 @@ def beaker_os_releases(at = nil)
         next unless setfile
 
         name = "#{puppet_version[:name]} - #{setfile[1]}"
+        # when we want to make this dynamic, switch to:
+        # "BEAKER_#{requirement_name.upcase}_COLLECTION"
```
I think the best way forward is to stick with `BEAKER_PUPPET_COLLECTION` ? If we actually stick to testing puppet and openvox packages in the long future, it makes sense to redo and rename beaker_puppet_helpers. But given how slow https://github.com/voxpupuli/community-triage/issues/29 makes progress, I doubt that.

ekohl commented:
When looking over the code it still didn't quite sit right with me that we hardcoded the requirements names in method names. In 0717bb817045fb40e14794e8cda58cdc03f942dd I took a stab at generalizing it. It's probably broken and for sure I didn't update the tests, but I wonder what you think of such an approach?

It also attempts to address the unit testing matrix where if the `puppet` requirement is missing that it doesn't generate anything. Instead I reuse the generalized method to just take the first requirement (which will be `openvox` if present) to generate the unit test matrix.

Once we start publishing the Puppet gem we should also come up with recommendations for `Gemfile`. For example, we may end up with:
```ruby
if ENV.key?('OPENVOX_VERSION')
  gem 'openvox', ENV['OPENVOX_VERSION']
elif ENV.key?('PUPPET_VERSION')
  gem 'puppet', ENV['puppet_version']
else
  gem 'openvox', '>= 7', '< 9'
end
```

I also thought about a Bundler plugin that reads `metadata.json` and generates a requirement based on that, but I doubt I'll have time to try that out.

bastelfreak commented:
@ekohl I cherry-picked your commits and had to adjust them slightly:

```diff
$ git diff 0717bb817045fb40e14794e8cda58cdc03f942dd..50c7387ee43d443b374bfa2a356df53ca8ed3cb1 -- lib/puppet_metadata/metadata.rb
diff --git a/lib/puppet_metadata/metadata.rb b/lib/puppet_metadata/metadata.rb
index 91c0964..7bb2181 100644
--- a/lib/puppet_metadata/metadata.rb
+++ b/lib/puppet_metadata/metadata.rb
@@ -140,12 +140,10 @@ module PuppetMetadata
     end

     def requirements_with_major_versions
-      SUPPORTED_REQUIREMENTS.to_h do |requirement|
+      SUPPORTED_REQUIREMENTS.filter_map do |requirement|
         majors = major_versions(requirement)
-        next if majors.empty?
-
-        [requirement, majors]
-      end
+        [requirement, majors] unless majors.empty?
+      end.to_h
     end

     # @return [Array[Integer]] Supported major Puppet versions
```

`Array.to_h` fails if one of the returned elements are nil and not and Array, so I switched to `.filter_map`. I also adjusted the tests and tried everything again in https://github.com/voxpupuli/puppet-example/pull/87.

rwaffen reviewed (approved):
LGTM

bastelfreak commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -49,17 +55,16 @@ def puppet_unit_test_matrix
     end
 
     def beaker_os_releases(at = nil)
-      majors = puppet_major_versions
```
done
